### ISSUE TITLE
Crash when converting stream to array: `Guarantee failed: [data.size() == 0]`

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#3767]
In `next`, ran this query (using the python client):

```py
r.range(1000).concat_map([r.row]).coerce_to('ARRAY').run(c)
```

Got this crash:
```cpp
Version: rethinkdb 1.16.0-1-95-g6f9f6d-dirty (debug) (GCC 4.6.3)
error: Error in ../src/rdb_protocol/datum_stream.cc at line 815:
error: Guarantee failed: [data.size() == 0]
error: Backtrace:
error: Wed Feb 11 16:33:20 2015

       1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:101
       2: backtrace_t::backtrace_t() at backtrace.cc:203
       3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4: format_backtrace(bool) at backtrace.cc:198
       5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6: ql::eager_datum_stream_t::accumulate_all(ql::env_t*, ql::eager_acc_t*) at datum_stream.cc:815
       7: ql::datum_stream_t::to_array(ql::env_t*) at datum_stream.cc:701
       8: ql::coerce_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at type_manip.cc:233
       9: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at op.cc:217
       10: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at term.cc:444
       11: ql::run(ql::protob_t<Query>, rdb_context_t*, signal_t*, ql::stream_cache_t*, ip_and_port_t const&, Response*) at term.cc:263
       12: rdb_query_server_t::run_query(ql::protob_t<Query> const&, Response*, signal_t*, client_context_t*, ip_and_port_t const&) at query_server.cc:62
```

**Comment by Tryneus** [comment]
This also exists on `v1.16.x`.

**Comment by Tryneus** [comment]
This appears to be because `r.range(...)` does not respect `batchspec_t::all()` by adding a `.with_at_most(500)` to the end.  This would suggest that accumulating any `r.range(...)` stream larger than 500 elements would fail.  The good news is that this is an `r_sanity_check`, so it won't crash the release server.

**Comment by VeXocide** [comment]
This was added as a performance optimisation since generating the first batch felt relatively slow. I'll take care of it by either simply removing it, or not applying it if we're in a `batchspec_t::all()`.

**Comment by Tryneus** [comment]
@VeXocide, sorry I actually already fixed this, just got interrupted before I could put it up for review.  Now in review 2595.

**Comment by Tryneus** [comment]
The fix has been approved and merged into `v1.16.x` in commit bcab24c, and cherry-picked into `next` in commit eec7134.  Will be in release 1.16.2.
